source/common/stats/thread_local_store.cc

```diff
@@ -699,26 +760,83 @@ void ThreadLocalHistogramImpl::merge(histogram_t* target) {
   hist_clear(*other_histogram);
 }
 
-ParentHistogramImpl::ParentHistogramImpl(StatName name, Histogram::Unit unit, Store& parent,
-                                         TlsScope& tls_scope, StatName tag_extracted_name,
+ParentHistogramImpl::ParentHistogramImpl(StatName name, Histogram::Unit unit,
+                                         ThreadLocalStoreImpl& parent, StatName tag_extracted_name,
                                          const StatNameTagVector& stat_name_tags,
-                                         ConstSupportedBuckets& supported_buckets)
+                                         ConstSupportedBuckets& supported_buckets, uint64_t id)
     : MetricImpl(name, tag_extracted_name, stat_name_tags, parent.symbolTable()), unit_(unit),
-      parent_(parent), tls_scope_(tls_scope), interval_histogram_(hist_alloc()),
-      cumulative_histogram_(hist_alloc()),
+      parent_(parent), interval_histogram_(hist_alloc()), cumulative_histogram_(hist_alloc()),
       interval_statistics_(interval_histogram_, supported_buckets),
-      cumulative_statistics_(cumulative_histogram_, supported_buckets), merged_(false) {}
+      cumulative_statistics_(cumulative_histogram_, supported_buckets), merged_(false), id_(id) {}
 
 ParentHistogramImpl::~ParentHistogramImpl() {
+  parent_.releaseHistogramCrossThread(id_);
+  ASSERT(ref_count_ == 0);
   MetricImpl::clear(parent_.symbolTable());
   hist_free(interval_histogram_);
   hist_free(cumulative_histogram_);
 }
 
+void ParentHistogramImpl::incRefCount() { ++ref_count_; }
+
+bool ParentHistogramImpl::decRefCount() {
+  bool ret;
+  if (shutting_down_) {
+    ret = --ref_count_ == 0;
+  } else {
+    ret = parent_.decHistogramRefCount(*this, ref_count_);
+  }
+  return ret;
+}
+
+bool ThreadLocalStoreImpl::decHistogramRefCount(ParentHistogramImpl& hist,
+                                                std::atomic<uint32_t>& ref_count) {
+  // We must hold the store's histogram lock when decrementing the
+  // refcount. Otherwise another thread may simultaneously try to allocate the
+  // same name'd stat after we decrement it, and we'll wind up with a
+  // dtor/update race. To avoid this we must hold the lock until the stat is
+  // removed from the map.
+  Thread::LockGuard lock(hist_mutex_);
+  ASSERT(ref_count >= 1);
+  if (--ref_count == 0) {
+    if (!shutting_down_) {
+      const size_t count = histogram_set_.erase(hist.statName());
+      ASSERT(shutting_down_ || count == 1);
+    }
+    return true;
+  }
+  return false;
+}
+
+uint32_t ThreadLocalStoreImpl::numHistograms() const {
```

Comment:
This seems like it's also for tests? Again wondering if this can be inferred by just calling the normal histograms() function to get all histograms? What does this do differently now that there are no dups?

Reply:
Good idea; can drop this.

Reply:
Good that I did; doing so caught a case where I wasn't removing late-rejected histograms from histogram_set_.